File: Algorithm/016. 3Sum Closest.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-03-10 17:38:42
# @Last modified by:   LC
# @Last Modified time: 2016-04-10 16:24:30
# @Email: liangchaowu5@gmail.com

class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        
        nums.sort()
        min = 1000000
        
        for i in range(len(nums)-2):
            j = i+1
            k = len(nums)-1
            while j<k:
                sum = nums[i]+nums[j]+nums[k]
                if sum > target:
                    gap = abs(sum -target)
                    if min>gap:
                        min = gap
                        result = sum
                    k-=1
                elif sum < target:
                    gap = abs(target - sum)
                    if min > gap:
                        min = gap
                        result = sum
                    j+=1
                else:
                    result = sum
                    return result
        return result
                
                    
                    

File: Algorithm/Java/016. 3Sum Closest.java
/**
* Author: WuLC
* Date:   2016-10-25 16:01:12
* Last modified by:   WuLC
* Last Modified time: 2016-10-25 16:13:01
* Email: liangchaowu5@gmail.com
*/


// sort and the use two pointers
// time complexity O(n^2)
public class Solution 
{
    public int threeSumClosest(int[] nums, int target) 
    {
        Arrays.sort(nums);
        Integer result = null, closest = null;
        int left, right, sum;
        for (int i=0; i<nums.length-2; i++)
        {
            left = i+1; 
            right= nums.length - 1; 
            sum = target - nums[i];
            closest = null;
            while (left < right)
            {
                int tmp = nums[left] + nums[right];
                if (tmp == sum) return target;
                else if(tmp > sum) right--;
                else left++;
                
                //update closest
                if (closest == null || Math.abs(closest-sum) > Math.abs(tmp-sum))
                    closest = tmp;
            }
        // update result
        if (result == null || Math.abs(closest+nums[i]-target) < Math.abs(result-target))
            result = closest+nums[i];
        }
        return result;
    }
    
}



// make the above code faster and more concise
public class Solution 
{
    public int threeSumClosest(int[] nums, int target) 
    {
        Arrays.sort(nums);
        int result = nums[0] + nums[1] + nums[2];
        int left, right, sum;
        for (int i=0; i<nums.length-2; i++)
        {
            left = i+1; 
            right= nums.length - 1; 
            while (left < right)
            {
                sum = nums[i] + nums[left] + nums[right];
                if (target == sum) return target;
                else if(target < sum) right--;
                else left++;
                
                if (Math.abs(result-target) > Math.abs(sum-target))
                    result = sum;
            }
        }
        return result;
    }
}